src/utils/communication_buffer.hpp

```diff
@@ -0,0 +1,418 @@
+      get_resource_(get_resource), buf_() {
+// Set up persistent communication
+#ifdef MPI_PARALLEL
+  PARTHENON_MPI_CHECK(MPI_Comm_rank(comm_, &my_rank));
+#else
+  my_rank = 0;
+#endif
+  if (send_rank == recv_rank) {
+    assert(my_rank == send_rank);
+    *comm_type_ = BuffCommType::both;
+  } else if (my_rank == send_rank) {
+    *comm_type_ = BuffCommType::sender;
+  } else if (my_rank == recv_rank) {
+    *comm_type_ = BuffCommType::receiver;
+    if (do_sparse_allocation) *comm_type_ = BuffCommType::sparse_receiver;
+  } else {
+    // This is an error
+    std::cout << "CommBuffer initialization error" << std::endl;
+  }
+}
+
+template <class T>
+template <class U>
+CommBuffer<T>::CommBuffer(const CommBuffer<U> &in)
+    : buf_(in.buf_), state_(in.state_), comm_type_(in.comm_type_),
+      started_irecv_(in.started_irecv_), nrecv_tries_(in.nrecv_tries_),
+      my_request_(in.my_request_), tag_(in.tag_), send_rank_(in.send_rank_),
+      recv_rank_(in.recv_rank_), comm_(in.comm_), active_(in.active_) {
+#ifdef MPI_PARALLEL
+  PARTHENON_MPI_CHECK(MPI_Comm_rank(comm_, &my_rank));
+#else
+  my_rank = 0;
+#endif
+}
+
+template <class T>
+CommBuffer<T>::~CommBuffer() {
+#ifdef MPI_PARALLEL
+  if (my_request_.use_count() == 1) { // This is the last shallow copy of this buffer
+    int flag;
+    MPI_Status status;
+    PARTHENON_MPI_CHECK(MPI_Test(my_request_.get(), &flag, &status));
+    if (!flag) {
+      if (*comm_type_ == BuffCommType::sender) {
+        PARTHENON_MPI_CHECK(MPI_Wait(my_request_.get(), MPI_STATUS_IGNORE));
+      } else {
+        PARTHENON_MPI_CHECK(MPI_Cancel(my_request_.get()));
+        PARTHENON_MPI_CHECK(MPI_Wait(my_request_.get(), MPI_STATUS_IGNORE));
+      }
+    }
+  }
+#endif
+}
+
+template <class T>
+template <class U>
+CommBuffer<T> &CommBuffer<T>::operator=(const CommBuffer<U> &in) {
+  buf_ = in.buf_;
+  state_ = in.state_;
+  comm_type_ = in.comm_type_;
+  started_irecv_ = in.started_irecv_;
+  nrecv_tries_ = in.nrecv_tries_;
+  my_request_ = in.my_request_;
+  tag_ = in.tag_;
+  send_rank_ = in.send_rank_;
+  recv_rank_ = in.recv_rank_;
+  comm_ = in.comm_;
+  active_ = in.active_;
+#ifdef MPI_PARALLEL
+  PARTHENON_MPI_CHECK(MPI_Comm_rank(comm_, &my_rank));
+#else
+  my_rank = 0;
+#endif
+  return *this;
+}
+
+template <class T>
+void CommBuffer<T>::Send() noexcept {
+  if (!active_) {
+    SendNull();
+    return;
+  }
+
+  PARTHENON_DEBUG_REQUIRE(*state_ == BufferState::stale,
+                          "Trying to send from buffer that hasn't been staled.");
+  *state_ = BufferState::sending;
+  if (*comm_type_ == BuffCommType::sender) {
+// Make sure that this request isn't still out,
+// this could be blocking
+#ifdef MPI_PARALLEL
+    PARTHENON_REQUIRE(
+        buf_.size() > 0,
+        "Trying to send zero size buffer, which will be interpreted as sending_null.");
+    PARTHENON_MPI_CHECK(MPI_Wait(my_request_.get(), MPI_STATUS_IGNORE));
```

Comment:
In which situation would this result in a wait?

Reply:
If the previous message sent from this buffer had not yet been received, which is possible if to ranks are substantially out of sink (i.e. one moves on to the next stage before the other is done receiving the previous stage).

Reply:
Can this actually happen?
IIRC we synchronize at TaskRegions (and within TaskLists). Given that all TaskRegions/Lists are defined within a loop where the counter is the stage in the driver.

If I'm not missing anything, this may actually be dangerous if we'd allow for this because all these comm vars are shared across stages so that accessing containers/`MeshData` based on the stage index could really mix things up.

Reply:
I don't think there are any `MPI_Barrier`s in the code, so I think it is possible for ranks to get out of sync even at the stage level. That being said, I think this check is extraneous since we are checking buffers with `IsAvailableForWrite()` previous to writing to the sending buffers. This should stop ranks from getting too far out of sync and having messages overrun each other.

Reply:
I am going to just leave this in since it is not doing anything bad I think